src/psyclone/psyir/backend/visitor.py

```diff
@@ -152,7 +152,35 @@ def __call__(self, node):
         :rtype: str
 
         '''
-        return self._visit(node)
+        if not isinstance(node, Node):
+            raise TypeError(
+                "The PSyIR visitor functor method only accepts a PSyIR Node "
+                "as argument, but found '{0}'.".format(type(node).__name__))
+
+        # The visitor must not have any side effect to the provided node, but
+        # if there are DSL concepts this will need to be lowered in-place and
+        # this operation often modifies the tree. Therefore, we first create a
+        # copy of the full provided tree (as modifications can be above the
+        # provided node - e.g. adding a symbol in the scope)
+        tree_copy = node.root.copy()
+
+        # Get the node in the new tree with equivalent position to the provided
+        # node
+        node_copy = tree_copy.walk(Node)[node.abs_position]
+
+        # Lower the DSL concepts starting from the selected node.
+        try:
+            node_copy.lower_to_language_level()
+        except Exception as error:
+            six.raise_from(VisitorError(
+                "Failed to lower '{0}'. Note that some nodes need to be "
+                "lowered from an ancestor in order to properly apply their "
+                "in-tree modifications.".format(node)), error)
+
+        # Find again the equivalent tree in case this has been replaced
```

Comment:
"...equivalent node in the lowered tree in case..."

Reply:
Done